### PR TITLE
Feature issue tokens

### PR DESCRIPTION
**Pull request by davikstone2**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,4 +2,5 @@
 target/
 dist/
 .DS_Store
-configs/config.json
\ No newline at end of file
+configs/config.json
+saito-js/
```
Comment (SankaD) [inline .gitignore:6]: remove this line

File: saito-core/Cargo.toml [hunk 2]
```diff
@@ -29,6 +29,7 @@ async-trait = "0.1.68"
 primitive-types = "0.12.1"
 futures = { version = "0.3.25" }
 log = "0.4.17"
+bs58 = "0.5.0"
```
Comment (SankaD) [inline saito-core/Cargo.toml:32]: why is this needed?

File: .gitignore (hunk 1 above)
Comment (SankaD) [inline .gitignore:6]: remove this line. shouldn't be in this workspace

File: saito-core/src/core/data/storage.rs [hunk 3]
```diff
@@ -174,6 +212,38 @@ impl Storage {
         return vec![];
     }
 
+    fn convert_issuance_into_slip(&self, line: &str) -> Slip {
+        let entries: Vec<&str> = line.split("\t").collect();
+        let amount = entries[0].parse::<u64>().expect("Failed to parse amount");
+        let publickey_str = entries[1];
+        let publickey_vec = bs58::decode(publickey_str)
```
Comment (SankaD) [inline saito-core/src/core/data/storage.rs:219]: better to have a utility function than directly using the lib here. (bs58)

File: saito-core/src/core/consensus_thread.rs [hunk 4]
```diff
@@ -138,6 +141,36 @@ impl ConsensusThread {
...
+        &self,
+        mempool: Arc<RwLock<Mempool>>,
+        blockchain: Arc<RwLock<Blockchain>>,
+    ) {
+        info!("generating issuance init transaction");
+
+        let (blockchain, _blockchain_) = lock_for_read!(blockchain, LOCK_ORDER_BLOCKCHAIN);
+        let (mut mempool, _mempool_) = lock_for_write!(mempool, LOCK_ORDER_MEMPOOL);
+
+        {
```
Comment (SankaD) [inline saito-core/src/core/consensus_thread.rs:154]: is this bracket needed? can remove

File: saito-core/src/core/consensus_thread.rs [hunk 5]
```diff
@@ -138,6 +141,36 @@ impl ConsensusThread {
...
         }
     }
+    async fn generate_issuance_tx(
+        &self,
+        mempool: Arc<RwLock<Mempool>>,
+        blockchain: Arc<RwLock<Blockchain>>,
+    ) {
+        info!("generating issuance init transaction");
+
+        let (blockchain, _blockchain_) = lock_for_read!(blockchain, LOCK_ORDER_BLOCKCHAIN);
```
Comment (SankaD) [inline saito-core/src/core/consensus_thread.rs:151]: these locks should be moved below. just before 'for tx in txs' line. critical sections should be as small as possible

File: saito-core/src/core/consensus_thread.rs [hunk 6]
```diff
@@ -520,6 +550,8 @@ impl ProcessEvent<ConsensusEvent> for ConsensusThread {
             .load_blocks_from_disk(self.mempool.clone())
             .await;
 
+        //
```
Comment (SankaD) [inline saito-core/src/core/consensus_thread.rs:553]: remove empty comment line

File: saito-core/src/core/data/storage.rs [hunk 7]
```diff
@@ -75,48 +78,55 @@ impl Storage {
 
     pub async fn load_blocks_from_disk(&mut self, mempool: Arc<RwLock<Mempool>>) {
         info!("loading blocks from disk");
+        self.get_token_supply_slips_from_disk().await;
```
Comment (SankaD) [inline saito-core/src/core/data/storage.rs:81]: why is this needed here? issuance related things shouldn't be in load_blocks

File: saito-core/src/core/data/storage.rs [hunk 8]
```diff
@@ -75,48 +78,55 @@ impl Storage {
...
             error!("failed loading blocks . {:?}", file_names.err().unwrap());
             return;
         }
+
         let mut file_names = file_names.unwrap();
-        file_names.sort();
-        debug!("block file names : {:?}", file_names);
+        if file_names.is_empty() {
+            // add slips
+            println!("disk is empty, adding slip")
```
Comment (SankaD) [inline saito-core/src/core/data/storage.rs:92]: why "adding slip" ?

File: saito-core/src/core/data/storage.rs [hunk 9]
```diff
@@ -136,10 +146,40 @@ impl Storage {
...
+        let mut v: Vec<Slip> = vec![];
+        let mut tokens_issued = 0;
         //
+        if self.file_exists(ISSUANCE_FILE_PATH).await {
+            if let Ok(lines) = self
+                .io_interface
+                .read_value(ISSUANCE_FILE_PATH.to_string())
+                .await
+            {
+                let mut contents = String::from_utf8(lines).expect("Failed to convert to String");
```
Comment (SankaD) [inline saito-core/src/core/data/storage.rs:159]: for IO related things check the returned result rather than .expect(). better to print the exact error also

File: saito-core/src/core/data/storage.rs [hunk 10]
```diff
@@ -136,10 +146,40 @@ impl Storage {
...
+
+                for i in 0..v.len() {
+                    tokens_issued += v[i].amount;
+                }
+
+                debug!("{:?} tokens issued", tokens_issued);
+                return v;
+            }
+        } else {
+            debug!("issuance file does not exist");
```
Comment (SankaD) [inline saito-core/src/core/data/storage.rs:180]: this should be an error. not debug

File: saito-core/src/core/data/storage.rs [hunk 11]
```diff
@@ -225,17 +298,17 @@ mod test {
     #[ignore]
     #[tokio::test]
     async fn read_issuance_file_test() {
-        let mut t = TestManager::new();
-        t.initialize(100, 100_000_000).await;
+        // let mut t = TestManager::new();
```
Comment (SankaD) [inline saito-core/src/core/data/storage.rs:301]: implement this test also

File: saito-rust/configs/issuance.template [hunk 12]
```diff
@@ -0,0 +1 @@
+100000000	03145c7e7644ab277482ba8801a515b8f1b62bcd7e4834a33258f438cd7e223849	VipOutput
```
Comment (SankaD) [inline saito-rust/configs/issuance.template:1]: this should be in the same place as the issuance file. (in data/issuance/. ?)

Comment (SankaD) [inline saito-rust/configs/issuance.template:1]: Is VipOutput a type or a label? if it's label, should be changed to 'Spammer'

File: saito-rust/src/main.rs [hunk 13]
```diff
@@ -100,6 +100,8 @@ where
                 .is_some()
             {
                 work_done = true;
+            } else {
+                // debug!("No work done")
```
Comment (SankaD) [inline saito-rust/src/main.rs:104]: remove else clause

**Review by SankaD — CHANGES_REQUESTED** [review]

File: saito-rust/configs/issuance.template (hunk 12 above)
Comment (davikstone2) [reply saito-rust/configs/issuance.template:1]: It is a type, There is no Spammer type in transactions.rs

File: saito-rust/data/issuance/issuance [hunk 14]
```diff
@@ -0,0 +1,4 @@
+5810123	n5hNA7Tazy7D226nq5arWq5YHxe8ibXGqNtjwXpEhnPT	VipOutput
```
Comment (SankaD) [inline saito-rust/data/issuance/issuance:1]: this file should not be in the repo.

File: saito-rust/.gitignore [hunk 15]
```diff
@@ -4,4 +4,3 @@ flamegraph.svg
 perf.data
 perf.data.old
 data/saito.stats
-data/issuance/issuance
```
Comment (SankaD) [inline saito-rust/.gitignore]: why was this removed?

File: saito-core/src/core/data/storage.rs [hunk 16]
```diff
@@ -177,7 +175,7 @@ impl Storage {
                 return v;
             }
         } else {
-            debug!("issuance file does not exist");
+            panic!("issuance file does not exist");
```
Comment (SankaD) [inline saito-core/src/core/data/storage.rs:175]: not panic. but error!(). panic!() will abort the process.

**Review by SankaD — CHANGES_REQUESTED** [review]

**Review by SankaD — CHANGES_REQUESTED** [review]

File: saito-core/src/core/data/storage.rs [hunk 17]
```diff
@@ -136,81 +139,80 @@ impl Storage {
...
-    //         slip.set_slip_type(SlipType::Normal);
-    //     }
-    //
-    //     return slip;
-    // }
+    fn convert_issuance_into_slip(&self, line: &str) -> Option<Slip> {
+        let entries: Vec<&str> = line.split("\t").collect();
+        let amount = entries[0]
+            .parse::<u64>()
+            .expect("Failed to parse amount from slip entry");
```
Comment (SankaD) [inline saito-core/src/core/data/storage.rs:184]: better print the exact error and some log and then terminate.

**Review by SankaD — APPROVED** [review]
@davikstone2 can merge this. please fix that error handling part in a separate PR.

**Review by SankaD — APPROVED** [review]
